### ISSUE TITLE
Note: state for selected row is kept on each row

### ISSUE THREAD

**Issue by krausest** [krausest/js-framework-benchmark#800]
Some frameworks can benefit from keeping the selected row or even active class for each row on the model for the row.
It's regarded to be good practice (and fair play) to have the selected row in your table model.
The impact for react and angular can be seen in PR #799

**Comment by hville** [comment]
What about imposing an external `Store` that can be imported like [@WebReflection did](github.com/WebReflection/js-framework-benchmark-utils)? 
Similar to a standard vendor API, it would level the sources of variances and also reduce some of the maintenance efforts by making all examples shorter.

**Comment by ryansolid** [comment]
Some libraries (all reactive libraries) use special primitives to do change detection. These are mutable generally, so non-proxy libraries examples would need to map the data before use.  If that was not done in the store you'd incur an additional mapping step. Unfortunately everyone of these libraries has their own primitives. 

Libraries like Svelte and I think Fidan use a compiler to mask these reactive primitives which has restrictions to where these occur. These syntaxes don't differentiate between reading a reactive atom and reading the value it holds always compiling to the latter, in so only properly working within the local scope of their components. They have their own approach to stores but they can bring additional overhead. In Svelte's case introducing subscription based reactivity instead of their usually component based. It's like forcing React to use Redux.

In short I'm not sure doing so let's most libraries play to their approaches. Many benchmarks take that approach and completely alienate reactive libraries.

**Comment by hville** [comment]
But ultimately all data has to come and go somewhere and that cost has to be born at least on every upstream exchange.

**Comment by hville** [comment]
But ultimately all data has to come and go somewhere and that cost has to be born at least on every upstream exchange. Not that I have a preference for this benchmark, it is all up to the goals and intent.

**Comment by ryansolid** [comment]
@hville Generally speaking unless the operation is a full diff you only need to pay this cost on initial creation/load. Unless you are in the habit of repeately loading large data snapshots this never comes up again. You set the data up (one time cost) and from there end-user updates don't need to pay the cost. It's what makes this benchmark much nicer to work with than like dbmon etc which don't even attempt to simulate end-user actions. 

If we prevent libraries that work this way from paying the cost upfront (which is their typical pattern) then they have to pay it on every interaction which basically negates their reason for existing since their creation cost is generally higher to optimize for updates. The whole purpose of these libraries is to trade CPU cost of Memory cost to improve update performance.

There are otherways for these sort of systems to achieve this with proxies but restricting it just forces those sort of solutions. While I'd say the majority of libraries don't work this way, there is still a significant number. I think it is probably just under 30% that do including Vue and Ember, among others. They have other means to deal with this but it can be a significant deoptimization.

**Comment by pzuraq** [comment]
I just wanted to echo @ryansolid's concerns here, this is not just about using a better structure to get ahead. In cases like Ember and Vue, structuring state like this is conventional and a fundamental part of the framework as a whole. The fact that state management is deeply integrated into the view layer, and that the view layer can optimize for state _without_ extra work done by the users of these frameworks, is a core part of their value proposition. The benefits users get from this include better performance out of the box. In fact, this exact issue is something we encounter frequently with new users of Ember, and the conventional advice is to restructure the state.

**Comment by ryansolid** [comment]
I think about this one often because it is an important distinction between select rows and update every 10th row. It also is a problem I have been trying to find the most general solution to for years since the time I worked for a company doing a lot of image management, with drag/drop, multi-page multi-select, etc.

Some people might find this classification harsh, but it is important to separate the underlying model from the ephemeral UI state. What if the model is used elsewhere with a different selection concept, you wouldn't want to pay the specific UI cost or deal with collisions.

So what are the options? As far as I can tell, there are 2 ways to solve this that should be considered fair for the benchmark:

1. Single selected state that is read/compared by each row. This is the common approach. It does N evals on each update, but is pretty common in re-render frameworks to have to re-run each component, ie using the UI's natural mapping to resolve.

2. Secondary mapped data structure. Like a hashmap based on id. There is lifecycle concern of course, because if you add 1000s of entries this can add up. It needs to hold up through moving rows/removal etc...

We haven't seen too much of the second solution. Solid's `createSelector` is an example but I think even more general solutions like self-releasing reactive proxy stores also fall into this category. As long as the you aren't polluting the main data model it is free game. I spent a lot of time coming up with ideas that would involve deriving these mapped structures from a signal state variable but realistically that isn't necessary to pass this check. I write this now because I suspect to see more reactive libraries take this approach in the near future because [we are generalizing this as a pattern](https://dev.to/playfulprogramming/mutable-derivations-in-reactivity-2ffl). And I want to sort of give a heads up.
